gax/src/main/java/com/google/api/gax/rpc/ServerStream.java

```diff
@@ -0,0 +1,110 @@
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalApi;
+import java.util.Iterator;
+import javax.annotation.Nonnull;
+
+/**
+ * A blocking Iterable-style wrapper around server stream responses.
+ *
+ * <p>This class asynchronously pulls responses from upstream via {@link
+ * StreamController#request(int)} and exposes them via its Iterator. The implementation is back
+ * pressure aware and uses a constant buffer of 1 item.
+ *
+ * <p>Please note that the stream can only be consumed once and must either be fully consumed or be
+ * canceled.
+ *
+ * <p>This class is not thread safe.
+ *
+ * <p>Example usage:
+ *
+ * <pre>{@code
+ * ServerStream<Item> stream = ...;
+ *
+ * for (Item item : stream) {
+ *   System.out.println(item.id());
+ *
+ *   // Allow for early termination
+ *   if (item.id().equals("needle")) {
+ *     stream.cancel();
+ *   }
+ * }
+ * }</pre>
+ *
+ * @param <V> The type of each response.
+ */
+@BetaApi("The surface for streaming is not stable yet and may change in the future.")
+public final class ServerStream<V> implements Iterable<V> {
+  private final QueuingResponseObserver<V> observer = new QueuingResponseObserver<>();
+  private final ServerStreamIterator<V> iterator = new ServerStreamIterator<>(observer);
+  private boolean consumed;
+
+  @InternalApi("For use by ServerStreamingCallable only.")
+  ServerStream() {}
+
+  @InternalApi("For use by ServerStreamingCallable only.")
+  ResponseObserver<V> observer() {
+    return observer;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  @Nonnull
+  public Iterator<V> iterator() {
+    if (consumed) {
+      throw new IllegalStateException("Iterator already consumed");
+    }
+    consumed = true;
```

Comment:
I think we should just make this class thread-safe (even if we don't guarantee it).
Calling `iterator()` feels like a read operation, and it's tempting to reason that calling `iterator()` concurrently is OK.

IIUC, this isn't hard to do:
```java
AtomicBoolean consumed = new AtomicBoolean();

if (consumed.getAndSet(true)) throw new IllegalStateException();
```

Reply:
It's easy to make ServerStream thread safe, but it's hard to make the actual Iterator thread safe.  I don't think there is value to be gained from making the outer getter thread safe, while leaving the actual iterator unsafe.

Reply:
Makes sense.